src/zjs_common.h

```diff
@@ -33,39 +33,77 @@
  */
 char *zjs_shorten_filepath(char *filepath);
 
-#ifdef DEBUG_BUILD
+// enable to use asserts even in release mode, e.g. for terse debugging
+#if 0
+#define USE_ASSERT
+#endif
 
+#ifdef DEBUG_BUILD
 int zjs_get_sec(void);
 int zjs_get_ms(void);
 
-#define DBG_PRINT                                                             \
+#define DBG_PRINT                                                       \
     ZJS_PRINT("\n%u.%3.3u %s:%d %s():\n(INFO) ", zjs_get_sec(), zjs_get_ms(), \
               zjs_shorten_filepath(__FILE__), __LINE__, __func__);            \
     ZJS_PRINT
-
 #define ERR_PRINT                                                              \
     ZJS_PRINT("\n%u.%3.3u %s:%d %s():\n(ERROR) ", zjs_get_sec(), zjs_get_ms(), \
               zjs_shorten_filepath(__FILE__), __LINE__, __func__);             \
     ZJS_PRINT
+#else  // !DEBUG_BUILD
+#define DBG_PRINT(fmt...) do {} while (0)
+#define ERR_PRINT                                                        \
+    ZJS_PRINT("\n%s:%d %s():\n(ERROR) ", zjs_shorten_filepath(__FILE__), \
+              __LINE__, __func__);                                       \
+    ZJS_PRINT
+#endif  // DEBUG_BUILD
 
+#if defined(DEBUG_BUILD) || defined(USE_ASSERT)
 #define ZJS_ASSERT(condition, str)                 \
     if (!(condition)) {                            \
         ERR_PRINT("ASSERTION FAILURE: %s\n", str); \
         do {} while (1);                           \
     }
+#else
+#define ZJS_ASSERT(condition, str) do {} while (0)
+#endif
 
-#else  // !DEBUG_BUILD
-
-#define DBG_PRINT(fmt...) do {} while (0);
-
-#define ERR_PRINT                                                        \
-    ZJS_PRINT("\n%s:%d %s():\n(ERROR) ", zjs_shorten_filepath(__FILE__), \
-              __LINE__, __func__);                                       \
+/**
+ * file-specific function tracing for debug purposes
+ *
+ * Displays short thread id, a file-specific prefix, and the function name,
+ *   followed by a formatted string intended to display arguments
+ *
+ * Use by defining USE_FTRACE and ftrace_prefix string in a file (see zjs_net.c)
+ *
+ * FTRACE takes format string and arguments like printf
+ * FTRACE_JSAPI takes not arguments and expects to run in a ZJS_DECL_FUNC API
+ */
+#ifdef USE_FTRACE
+#define FTRACE                                                   \
+    ZJS_PRINT("[%x] %s: %s: ", (u32_t)k_current_get() & 0xffff,  \
+              FTRACE_PREFIX, __func__);                          \
     ZJS_PRINT
+#define FTRACE_JSAPI                                             \
```

Comment:
Could this just move into ZJS_DECL_FUNC? Then you it would be in all modules without having to call it everywhere.

Reply:
Not the way ZJS_DECL_FUNC is done currently because it doesn't include the opening brace of the function, so you can't add statements to the function. It could perhaps go into ZJS_VALIDATE_ARGS though.